### DISCUSSION TITLE
table placement quarto-pdf

### DISCUSSION THREAD

**Discussion by bernardo-cruz** [quarto-dev/quarto-cli#4849] [Q&A] [answered]
I am trying to write my thesis using quarto. I have created a table as follows:

```{python}
#| label: tbl-ohe_example
#| tbl-cap: One-hot vectorization example
#| tbl-pos: H

from IPython.display import display, Markdown, Latex, HTML, Image

Markdown(
    pd.DataFrame(
        data = {
            'Sentence': ['I like to eat apples', 'I like to eat bananas'],
            'Tokens' : [['I', 'like', 'to', 'eat', 'apples'], ['I', 'like', 'to', 'eat', 'bananas']],
            'Vector' : [[1, 1, 1, 1, 1, 0], [1, 1, 1, 1, 0, 1]],
        }
    )
   .to_html(index=False)
)
```
The table is perfectly rendered in the pdf - but placed in the top of the specific page of the document. The option `tbl-pos: H` seems not to function as `fig-pos: H` would. 

I even tried to import `\usepackage{float}` - see yaml header:

```
---
...
    pdf:  
      include-in-header: 
        text: |
          \usepackage{float}
      papersize: a4paper
      df-print: paged
      fig-pos: 'H'
      output-file: "thesis"
      output-ext: "pdf"
...
---
```

Any hints how to avoid floating tables?

**Comment by rgaiacs** [comment, marked as answer]
@bernardo-cruz Can you add `keep-tex: true` to your `_quarto.yaml` file? You will have something like

```
---
...
    pdf:  
      papersize: a4paper
      keep-tex: true
      df-print: paged
      fig-pos: 'H'
      output-file: "thesis"
      output-ext: "pdf"
...
---
```

Inspect the `.tex` file will help to diagnose the problem as the LaTeX engine has a lot of rules for placement of float elements.

**Reply by bernardo-cruz** [reply]
@rgaiacs THX very much for your answer! ☺️

what should I look for in the .text output? 
I see know the package `\usepackage{float}` being imported, but no option of `H` is passed to the table - see extract below:

```
\hypertarget{tbl-ohe_example}{}
\begin{longtable}[]{@{}lll@{}}
\caption{\label{tbl-ohe_example}One-hot vectorization
example}\tabularnewline
\toprule()
Sentence & Tokens & Vector \\
\midrule()
\endfirsthead
\toprule()
Sentence & Tokens & Vector \\
\midrule()
\endhead
I like to eat apples & {[}\textquotesingle I\textquotesingle,
\textquotesingle like\textquotesingle,
\textquotesingle to\textquotesingle,
\textquotesingle eat\textquotesingle,
\textquotesingle apples\textquotesingle{]} & {[}1, 1, 1, 1, 1, 0{]} \\
I like to eat bananas & {[}\textquotesingle I\textquotesingle,
\textquotesingle like\textquotesingle,
\textquotesingle to\textquotesingle,
\textquotesingle eat\textquotesingle,
\textquotesingle bananas\textquotesingle{]} & {[}1, 1, 1, 1, 0, 1{]} \\
\bottomrule()
\end{longtable}
```
you don't know how to pass the `H` argument? 

thx very much

**Reply by rgaiacs** [reply]
From https://en.wikibooks.org/wiki/LaTeX/Tables#Floating_with_table, the vanilla LaTeX code for float table is

```
\begin{table}[position specifier]
  \centering
  \begin{tabular}{|l|}
    ... your table ...
  \end{tabular}
  \caption{This table shows some data}
  \label{tab:myfirsttable}
\end{table}
```

and you can decide on the table position.

You are using `longtable` and I can't find in the [documentation](https://packages.oth-regensburg.de/ctan/macros/latex/required/tools/longtable.pdf) how to decide on the table position.

**Reply by cscheid** [reply]
Just to be clear, `longtable` is [currently hard-coded by Pandoc](https://github.com/jgm/pandoc/issues/5557).

**Reply by rgaiacs** [reply]
I forgot that @bernardo-cruz is using Python. I remember using R Markdown and knitr producing different LaTeX code for tables.

@bernardo-cruz if you are only producing PDF with LaTeX, quarto allows you to use raw LaTeX in your document, see https://quarto.org/docs/output-formats/pdf-basics.html#raw-latex.

Try change your code from

```
Markdown(
    pd.DataFrame(
        data = {
            'Sentence': ['I like to eat apples', 'I like to eat bananas'],
            'Tokens' : [['I', 'like', 'to', 'eat', 'apples'], ['I', 'like', 'to', 'eat', 'bananas']],
            'Vector' : [[1, 1, 1, 1, 1, 0], [1, 1, 1, 1, 0, 1]],
        }
    )
   .to_html(index=False)
)
```

to something like

```
pd.DataFrame(
    data = {
        'Sentence': ['I like to eat apples', 'I like to eat bananas'],
        'Tokens' : [['I', 'like', 'to', 'eat', 'apples'], ['I', 'like', 'to', 'eat', 'bananas']],
        'Vector' : [[1, 1, 1, 1, 1, 0], [1, 1, 1, 1, 0, 1]],
    }
)
.to_latex(index=False)
```

**Reply by bernardo-cruz** [reply]
@cscheid @rgaiacs thx very much for your answers - I am really grateful. 

In summary, I cannot use the default quarto setup for python (pandas) tables, right? I understand the point about `longtable` etc, but I don't understand how to change that since I use quarto to create pandoc and pdfs. Where can I define if a table is `longtable` or not? 

the code below yields to not output in the pdf
```
pd.DataFrame(
    data = {
        'Sentence': ['I like to eat apples', 'I like to eat bananas'],
        'Tokens' : [['I', 'like', 'to', 'eat', 'apples'], ['I', 'like', 'to', 'eat', 'bananas']],
        'Vector' : [[1, 1, 1, 1, 1, 0], [1, 1, 1, 1, 0, 1]],
    }
)
.to_latex(index=False)
```

**Reply by cscheid** [reply]
You need this:

````
---
format: latex
---

```{python}
import pandas as pd
```

```{python}
#| output: asis
#| warning: false
print(pd.DataFrame(
    data = {
        'Sentence': ['I like to eat apples', 'I like to eat bananas'],
        'Tokens' : [['I', 'like', 'to', 'eat', 'apples'], ['I', 'like', 'to', 'eat', 'bananas']],
        'Vector' : [[1, 1, 1, 1, 1, 0], [1, 1, 1, 1, 0, 1]],
    }
).to_latex(index=False))
```
````

**Reply by bernardo-cruz** [reply]
Thank you it works, but now the table is overflowing. Is there a reason why you have defined the `format: latex`? 
My intention was to write a pdf document.

**Reply by cscheid** [reply]
> Thank you it works, but now the table is overflowing.

You'll need to solve that by learning the pandas `to_latex()` API. Because of `output: asis`, quarto is not doing anything with that table. It's being emitted directly into the .tex source.

> Is there a reason why you have defined the `format: latex`

Oh, sorry. The only difference between `format: latex` and `format: pdf` is whether we stop at the .tex file or we generate a .pdf from it.

**Reply by bernardo-cruz** [reply]
@cscheid Thx very much for all your explanations! 👍

**Comment by Jonasrg** [comment]
I tried the above answer suggested by @cscheid. The problem was that while it placed the table correctly, Quarto no longer added the table headings and labels. I found a workaround though. Given the following code, Quarto places the table correctly within the text (not on top of the page) and gives it the appropiate caption and label.

```{python}
#| label: tbl-mytable
#| tbl-cap: This is awesome
#| tbl-pos: H

from IPython.display import display, Latex

display(Latex(df.to_latex(index=False)))
```
Hope that helps.

**Reply by awcm0n** [reply]
After hours of trying to fix this issue, `#| tbl-pos: H` caused the table to be placed correctly.

**Reply by math-mcshane** [reply]
FWIW, I was not able to replicate with `#| tbl-pos: H` in Quarto BUT I was able to add this to the YAML header and get the tables to stay in fixed position: 

```
    include-in-header:
      text: |
        \usepackage{float}
        \floatplacement{table}{H}
```

**Reply by mcanouil** [reply]
I don't know what "I was not able to replicate" means but `tbl-pos` works at fenced div and code cell level for now.

**Reply by math-mcshane** [reply]
@mcanouil  I was not able to replicate the desired result of fixed table placement -- the option did not work in multiple code cells outputting a `gt` table.

**Reply by mcanouil** [reply]
Please provide a small fully reproducible example and please ensure you are using the latest Quarto version.

Last time I checked (yesterday) the option worked at the two level I mentioned.

**Reply by raffaelemancuso** [reply]
Can I ask where `tbl-pos` is documented? Because it's not documented [here](https://quarto.org/docs/authoring/tables.html) (it solved the issue for me ... what luck that I had this issue today and found this post of yesterday!)

**Reply by mcanouil** [reply]
- https://github.com/quarto-dev/quarto-cli/issues/10784

**Reply by grantmcdermott** [reply]
Relatedly: has anyone managed to make this work in a raw markdown table div? E.g.

```quarto
See @tbl-ex for a demo.

| Default | Left | Right | Center |
|---------|:-----|------:|:------:|
| 12      | 12   |    12 |   12   |
| 123     | 123  |   123 |  123   |
| 1       | 1    |     1 |   1    |

: Demonstration of pipe table syntax {#tbl-ex tbl-pos="t"}
```

Doesn't work for me (the table still appears underneath the text rather than the top of the page.)

**Reply by mcanouil** [reply]
Did you look at the LaTeX code? `keep-tex: true`

**Reply by mcanouil** [reply]
I can't check right now but here is a test/example from the codebase:

https://github.com/quarto-dev/quarto-cli/blob/d86da0b118220ab8b9aeeb98e831abfc279b4073/tests/docs/smoke-all/2024/10/23/issue-10001.qmd

It might be a longtable (Pandoc) / table issue.

**Reply by mcanouil** [reply]
It seems `tbl-pos` is only respected when using layout.

- `tbl-pos="t"` is not respected

````qmd
---
format: pdf
keep-tex: true
---

{{< lipsum 1 >}}


::: {#tbl-tm2 tbl-pos="t"}

| A | B | C |
|---|---|---|
| 1 | 2 | 3 |

Not so nice: Where Latex Wants it
:::
````

- `tbl-pos="t"` is respected

````qmd
---
format: pdf
keep-tex: true
---

{{< lipsum 1 >}}


::: {#tbl-tm2 tbl-pos="t" layout-ncol="1"}

| A | B | C |
|---|---|---|
| 1 | 2 | 3 |

Not so nice: Where Latex Wants it
:::
````

That is because the table is actually encapsulated in a table/minipage via Quarto cross-reference div syntax.
This does not work via Pandoc syntax.

**Reply by grantmcdermott** [reply]
Thanks @mcanouil (you are a scholar and gent). I was slammed with deadlines at work yesterday, so couldn't test myself. I appreciate the `layout` workaround.
